src/distributed_planner/distributed_physical_optimizer_rule.rs

```
@@ -89,86 +90,174 @@ impl PhysicalOptimizerRule for DistributedPhysicalOptimizerRule {
-
-            return Ok(Transformed::yes(Arc::new(node)));
         }
+    }
 
-        // If this is a CoalescePartitionsExec, it means that the original plan is trying to
-        // merge all partitions into one. We need to go one step ahead and also merge all tasks
-        // into one.
-        if let (Some(node), Some(input_tasks)) = (
-            plan.as_any().downcast_ref::<CoalescePartitionsExec>(),
-            cfg.network_coalesce_tasks,
-        ) {
-            // If the immediate child is a PartitionIsolatorExec, it means that the rest of the
-            // plan is just a couple of non-computational nodes that are probably not worth
-            // distributing.
-            if node.input().as_any().is::<PartitionIsolatorExec>() {
-                return Ok(Transformed::no(plan));
-            }
+    let d_cfg = DistributedConfig::from_config_options(cfg)?;
+    let Some(mut ctx) = ctx else {
+        // As there is no context, it means that children.is_empty() == true and no ctx was set, so
+        // this is a leaf node, maybe a DataSourceExec, or maybe something else custom from the
+        // user. We need to estimate how many tasks are needed for this leaf node, as we'll use
+        // that for choosing the amount of tasks in upper stages.
+        let Some(estimate) = d_cfg.__private_task_estimator.estimate_tasks(&plan, cfg) else {
+            // We could not determine how many tasks this leaf node should run on, so
+            // assume it cannot be distributed and employ just 1 task.
+            return Ok(ApplyNetworkBoundariesCtx {
+                task_count: 1,
+                this_stage_sf: 1.0,
+                next_stage_sf: 1.0,
+                plan,
+            });
+        };
+        return Ok(ApplyNetworkBoundariesCtx {
+            task_count: estimate.task_count,
+            this_stage_sf: 1.0,
+            next_stage_sf: 1.0,
+            plan: estimate.new_plan.unwrap_or(plan),
```

We should probably bubble this error up. It's bad if a user returned a task estimate but the node didn't transform. It could cause read amplification if there's 10 tasks for a DataSourceExec but the PartitionIsolator was not applied.

---

The `unwrap_or` here is not really unwrapping an error (`Result<>`) it's unwrapping an `Option<>`. It's fine if the user returned a `TaskEstimation` with `new_plan == None`, as long as they handle `DistributedTaskContext` in their own custom nodes.

Remember that `PartitionIsolatorExec` here is relevant for `DataSourceExec(FileScanConfig)` nodes, but it is not necessarily relevant for anything else, therefor returning `new_plan == None` is completely fine.

---

True. This is fine then!